solar_consumer/save/save_data_platform.py

```diff
@@ -0,0 +1,173 @@
+    """
+
+    assert "target_datetime_utc" in data_df.columns
+    assert "solar_generation_kw" in data_df.columns
+    assert "gsp_id" in data_df.columns
+    assert "regime" in data_df.columns
+    assert "capacity_mwp" in data_df.columns
+
+    # 1. Get all locations (UK GSPs and National)
+    all_locations = await get_all_gsp_and_national_locations(client)
+
+    # 2. Create an observer for that gsp and regime if it doesn't already exist
+    regime = data_df["regime"].unique()
+    assert len(regime) == 1, "DataFrame must contain only one regime type"
+    regime = regime[0]
+    name = f"PVLive_consumer_{regime}".lower().replace("-", "_")
+
+    list_observers_request = dp.ListObserversRequest(observer_names_filter=[name])
+    list_observers_response = await client.list_observers(list_observers_request)
+    if len(list_observers_response.observers) == 0:
+        observer_request = dp.CreateObserverRequest(name=name)
+        _ = await client.create_observer(observer_request)
+
+    # for each gsp
+    gsp_ids = data_df["gsp_id"].unique()
+    for gsp_id in gsp_ids:
+        logger.info(f"Saving GSP ID: {gsp_id} to Data Platform")
+
+        # 3. Get only the data for that gsp
+        gsp_data = data_df[data_df["gsp_id"] == gsp_id]
+
+        # 4. Get the location for that gsp
+        location = all_locations.get(gsp_id)
+        if location is None:
+            logger.warning(f"No location found for GSP ID {gsp_id}, skipping.")
+            continue
+        location_uuid = location.location_uuid
+
+        # 5. Get the most recent observations for that location and observer
+        try:
+            recent_observations_request = dp.GetLatestObservationRequest(
+                location_uuid=location_uuid,
+                energy_source=dp.EnergySource.SOLAR,
+                observer_name=name,
+            )
+            recent_observation_response = await client.get_latest_observation(
+                recent_observations_request
+            )
+
+            # 6. Remove any data points from our data that are already in the database
+            last_datetime_utc = recent_observation_response.timestamp_utc
+            gsp_data = gsp_data[
+                gsp_data["target_datetime_utc"] > last_datetime_utc
+            ]
+        except Exception:
+            logger.info(
+                f"No existing observations for location {location_uuid} and observer {name}, adding all data."
+            )
+
+
+        if len(gsp_data) == 0:
+            logger.debug(
+                f"No new data to add for location {location_uuid} and observer {name}, for gsp id {gsp_id}."
+            )
+            continue
+
+        # 7. Update location capacity based on the max capacity in this data
+        new_max_capacity_watts = int(gsp_data["capacity_mwp"].max() * 1_000_000)
+        max_capacity_watts_current = location.effective_capacity_watts
+        if new_max_capacity_watts > max_capacity_watts_current:
```

Comment:
@dfulu has pointed out that the capacities also might change to become less than before, and we should track that as well - so maybe worth doing a != here (although this whole thing might get folded into the dataframe operations anyway!)

Reply:
Oh good point

Reply:
done, thanks @dfulu